sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala

```diff
@@ -1642,4 +1869,152 @@ case class GpuHashAggregateExec(
   //
   // End copies from HashAggregateExec
   //
+
+  // We are not going to override requiredChildOrdering because we don't want to
+  // always sort the data. So we will insert in the sort ourselves if we need to
+  def expectedChildOrderingIfNeeded: Seq[SortOrder] = {
+    groupingExpressions.map(SortOrder(_, Ascending))
+  }
 }
+
+class DynamicGpuPartialSortAggregateIterator(
+    cbIter: Iterator[ColumnarBatch],
+    inputAttrs: Seq[Attribute],
+    groupingExprs: Seq[NamedExpression],
+    boundGroupExprs: GpuTieredProject,
+    aggregateExprs: Seq[GpuAggregateExpression],
+    aggregateAttrs: Seq[Attribute],
+    resultExprs: Seq[NamedExpression],
+    modeInfo: AggregateModeInfo,
+    estimatedPreGrowth: Double,
+    alreadySorted: Boolean,
+    ordering: Seq[SortOrder],
+    postBoundReferences: BoundExpressionsModeAggregates,
+    configuredTargetBatchSize: Long,
+    metrics: GpuHashAggregateMetrics,
+    useTiered: Boolean,
+    forceSinglePassAgg: Boolean,
+    allowSinglePassAgg: Boolean) extends Iterator[ColumnarBatch] {
+  private var aggIter: Option[Iterator[ColumnarBatch]] = None
+  private[this] val isReductionOnly = boundGroupExprs.outputTypes.isEmpty
+
+  // When doing a reduction we don't have the aggIter setup for the very first time
+  // so we have to match what happens for the normal reduction operations.
+  override def hasNext: Boolean = aggIter.map(_.hasNext)
+      .getOrElse(isReductionOnly || cbIter.hasNext)
+
+  private[this] def estimateCardinality(cb: ColumnarBatch): Int = {
+    withResource(boundGroupExprs.project(cb)) { groupingKeys =>
+      withResource(GpuColumnVector.from(groupingKeys)) { table =>
+        table.distinctCount()
+      }
+    }
+  }
+
+  private[this] def firstBatchHeuristic(
+      cbIter: Iterator[ColumnarBatch],
+      helper: AggHelper): (Iterator[ColumnarBatch], Boolean) = {
+    // we need to decide if we are going to sort the data or not, so the very
+    // first thing we need to do is get a batch and make a choice.
+    withResource(new NvtxWithMetrics("dynamic sort heuristic", NvtxColor.BLUE,
+      metrics.opTime, metrics.heuristicTime)) { _ =>
+      val cb = cbIter.next()
+      lazy val estimatedGrowthAfterAgg: Double = closeOnExcept(cb) { cb =>
+        val numRows = cb.numRows()
+        val cardinality = estimateCardinality(cb)
+        val minPreGrowth = PreProjectSplitIterator.calcMinOutputSize(cb,
+          helper.preStepBound).toDouble / GpuColumnVector.getTotalDeviceMemoryUsed(cb)
+        (math.max(minPreGrowth, estimatedPreGrowth) * cardinality) / numRows
+      }
+      val wrappedIter = Seq(cb).toIterator ++ cbIter
+      (wrappedIter, estimatedGrowthAfterAgg > 1.0)
```

Comment:
is 1.0 too small as a threshold? currently a simple query like below will incur single pass:

```
spark.time(spark.range(0, 1000000000L, 1, 2).selectExpr("id + 1 as key", "id as value").groupBy("key").agg(avg(col("value")).alias("avg_v")).orderBy(desc("avg_v"), col("key")).show())
```

The reason is simply because avg aggregate will use two buffers (count and sum), so the total size after aggregate is 1.5x of the input.

Since you mentioned `This is my first PR for an optimization for aggregations where there are a lot of aggregations` , obviously the example query can't be counted as `a lot of aggregations`

Reply:
The heuristic is simply looking at the cost of sorting now vs sorting later. 1.0 is the correct cutoff for that decision. The assumption that a sort will happen later is what is not correct. Nor is the assumption that a single batch is representative of the entire task's data. But we don't have the information we need to know if that sort is going to happen or not. We don't know how much data is coming later on. This could be the only batch in which case the sort is useless. There could be 100 more batches coming. If that is true, then sorting the data now will be faster than sorting it later on because there is less data to spill (even if it is just a little less data). Also we don't have a good way to start sorting later after we have already done part of the operation. 

If we want to fix this we really want to fix all of aggregation and have unified heuristics to decide what to do an when.

https://github.com/NVIDIA/spark-rapids/issues/11042

describes a starting point of moving to HyperLogLog for doing some operations. That includes updating this code to do an approximate count distinct instead of a full on count distinct (assuming the performance is good enough).

We also have https://github.com/NVIDIA/spark-rapids/issues/8622 which is to replace the sort here with a hash re-partition operation #11042 also talks about doing hash re-partitioning instead of sorting.

To me what we want is a somewhat lazy aggregation operation. I'll update #11042 to reflect some ideas I have for this.